File: server/network.go
package main

import (
	"encoding/json"
	"fmt"
	"net"

	"github.com/notnil/chess"
)

const BufferSize = 1024

// actions
const (
	StartGame      = "sg" // start code to send to clients when 2 players were found
	RequestingGame = "rg" // client connected and is requesting a game
	Ready          = "r"  // ready check for clients
	SendMove       = "sm" // sending a move
	UpdateBoard    = "ub" // sending a new updated board state
)

// represents a message between client and server
type Message struct {
	Action string                 `json:"action"`
	From   net.Conn               `json:"-"`
	Data   map[string]interface{} `json:"data"`
}

// returns json bytes for a Message
func messageJSON(m *Message) []byte {
	msgJSON, _ := json.Marshal(m)
	return msgJSON
}

// converts chess.board's squaremap to sendable data
func boardData(b chess.Board) map[string]interface{} {
	boardData := make(map[string]interface{})

	for square, piece := range b.SquareMap() {
		boardData[square.String()] = piece.Color().String() + piece.Type().String()
	}

	return boardData
}

// sends a start message with initial board state and players assigned side to the player
func sendStart(playerConn net.Conn, color chess.Color, initialBoard chess.Board) {
	data := map[string]interface{}{
		"color": color.String(),
		"board": boardData(initialBoard),
	}

	msg := Message{Action: StartGame, Data: data}
	playerConn.Write(messageJSON(&msg))
}

// send a new board state to both players
func sendBoard(player1 net.Conn, player2 net.Conn, board chess.Board) {

	msg := Message{Action: UpdateBoard, Data: boardData(board)}
	player1.Write(messageJSON(&msg))
	player2.Write(messageJSON(&msg))
}

// wrapper for making a buffer and reading a message from connection
func readFromConn(conn net.Conn) (Message, error) {
	buffer := make([]byte, BufferSize)
	len, err := conn.Read(buffer)

	trimmedMsg := buffer[:len]

	var response Message
	json.Unmarshal(trimmedMsg, &response)
	response.From = conn

	return response, err
}

// writes a ready message to the connection and expects one back
func connIsAlive(conn net.Conn) bool {
	msg := Message{Action: Ready, Data: map[string]interface{}{}}
	conn.Write(messageJSON(&msg))

	response, _ := readFromConn(conn)
	return response.Action == Ready
}

// continously gets messages from a connection and puts them in msgChan
func getMessages(playerConn net.Conn, msgChan chan<- Message) {
	for {
		msg, err := readFromConn(playerConn)
		if err != nil {
			fmt.Println(err)
		}
		msgChan <- msg
	}
}

File: client/network.py
import socket
import json
import queue

from enum import Enum
from dataclasses import dataclass

from thread import threaded

BUFFER_SIZE = 1024
ENCODING = "utf-8"


# possible actions to communicate with the server
class Action(str, Enum):
    START_GAME = "sg"
    REQUEST_GAME = "rg"
    READY = "r"
    SEND_MOVE = "sm"
    UPDATE_BOARD = "ub"
    CONCLUDE = "c"


# represents a message sent between client and server
@dataclass
class Message:
    action: Action
    data: dict
    
    # returns a message object created from raw json bytes
    @staticmethod
    def from_json_bytes(json_bytes: bytes):
        msg_json = json.loads(json_bytes)
        return Message(msg_json["action"], msg_json["data"])
    
    # returns the message represented as json bytes
    def to_json_bytes(self):
        payload = {
            "action": self.action,
            "data": self.data
        }

        json_str = json.dumps(payload)
        return bytes(json_str, ENCODING)


class Client:

    def __init__(self, server_address: str, server_port: str):
        self.server_address = server_address
        self.server_port = server_port

        self.messages = queue.Queue()
        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

        self.color = None

    # connect the socket
    def connect(self):
        self.socket.connect((self.server_address, self.server_port))
    
    # read one message from the socket
    def get_message(self):
        resp_bytes = self.socket.recv(BUFFER_SIZE)
        return Message.from_json_bytes(resp_bytes)
    
    # send a message's json bytes to the socket
    def send_message(self, msg: Message):
        self.socket.send(msg.to_json_bytes())

    # sends a start game message to the server
    # side assigned by the server will be set to self.color
    @threaded
    def request_game(self):
        msg = Message(Action.REQUEST_GAME, {})
        self.send_message(msg)
        
        opponent_found = False

        while not opponent_found:
            response = self.get_message()

            # send a ready message if the server is checking 
            if response.action == Action.READY:
                self.send_message(Message(Action.READY, {}))

            # opponent was found and game is starting
            elif response.action == Action.START_GAME:
                self.color = response.data["color"]
                self.messages.put(response)
                opponent_found = True

    # listen for incoming communication, puts them in self.messages queue
    @threaded
    def listen(self):
        while True:
            msg = Message.from_json_bytes(self.socket.recv(BUFFER_SIZE))
            self.messages.put(msg)

    # send a move message
    @threaded
    def send_move(self, from_square: str, to_square: str, promotion_choice=None):
        data = {"move": f"{from_square}{to_square}"}
        if promotion_choice:
            data["move"] += promotion_choice
        
        self.send_message(Message(Action.SEND_MOVE, data))
